aea/aea_builder.py

```diff
@@ -0,0 +1,746 @@
+        :return: the AEA object.
+        """
+        wallet = Wallet(self.private_key_paths)
+        identity = self._build_identity_from_wallet(wallet)
+        connections = self._process_connection_ids(connection_ids)
+        for connection in connections:
+            connection.address = identity.address
+            connection.load()
+        aea = AEA(
+            identity,
+            connections,
+            wallet,
+            LedgerApis(self.ledger_apis_config, self._default_ledger),
+            self._resources,
+            loop=None,
+            timeout=0.0,
+            is_debug=False,
+            is_programmatic=True,
+            max_reactions=20,
+        )
+        self._set_agent_context_to_all_skills(aea.context)
+        return aea
+
+    def _set_agent_context_to_all_skills(self, context: AgentContext):
+        """Set a skill context to all skills"""
+        for skill in self._resources.get_all_skills():
+            logger_name = "aea.{}.skills.{}.{}".format(
+                context.agent_name, skill.configuration.author, skill.configuration.name
+            )
+            skill.skill_context.set_agent_context(context)
+            skill.skill_context._logger = logging.getLogger(logger_name)
+
+    def _check_configuration_not_already_added(self, configuration):
+        if (
+            configuration.component_id
+            in self._package_dependency_manager.all_dependencies
+        ):
+            raise ValueError(
+                "Component {} of type {} already added.".format(
+                    configuration.public_id, configuration.component_type
+                )
+            )
+
+    def _check_package_dependencies(self, configuration):
+        self._package_dependency_manager.check_package_dependencies(configuration)
+
+    @staticmethod
+    def _find_component_directory_from_component_id(
+        aea_project_directory: Path, component_id: ComponentId
+    ):
+        """Find a component directory from component id."""
+        # search in vendor first
+        vendor_package_path = (
+            aea_project_directory
+            / "vendor"
+            / component_id.public_id.author
+            / component_id.component_type.to_plural()
+            / component_id.public_id.name
+        )
+        if vendor_package_path.exists() and vendor_package_path.is_dir():
+            return vendor_package_path
+
+        # search in custom packages.
+        custom_package_path = (
+            aea_project_directory
+            / component_id.component_type.to_plural()
+            / component_id.public_id.name
+        )
+        if custom_package_path.exists() and custom_package_path.is_dir():
+            return custom_package_path
+
+        raise ValueError("Package {} not found.".format(component_id))
+
+    @staticmethod
+    def _try_to_load_agent_configuration_file(aea_project_path: Path):
+        """Try to load the agent configuration file.."""
```

Comment:
this is the same code in the `aea.cli.helpers` module

Reply:
it could be refactored eventually.

Reply:
I'm thinking that `@check_aea_project` in `aea.cli` might be implemented with the builder, and if the builder manages to add all the components then it means the project is consistent. 

However, that would raise the following issue: should the component be loaded at "add" time (that is, when `add_component` is called) or at `build` time, before instantiating the `AEA` object?
Currently, we do that at "add" time, so the client of this class gets notified soon if he tried to add a bad component. However, that's unpractical if we want to use it in place of `check_aea_project`, because every command would need to load all the packages.

A solution would be to try to load the component at "add" time, but without instantiating a component object.